### PR TITLE
feat: Production Planning Against Sales Order/Material Request/Work Order

### PR DESCRIPTION
**Pull request by rohitwaghchaure**
**Production Planning For Sales Order**
image

**Production Planning For Material Request (Type as Manufacture)**
image

**Production Planning For Work Order**
image

### CODE REVIEW

**Comment by djpalshikar** [comment]
@rohitwaghchaure thanks for the report :)

**Comment by toykraft** [comment]
Rohit, thank you for the report... please confirm the following:
does the report show only work orders that are active or all? 

Also, can we have a reverse report as well, where a raw material which could be missing in many items ... this will trigger a production when that raw material is received!

File: erpnext/manufacturing/report/production_planning_report/production_planning_report.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+			"fieldname": "required_qty",
+			"fieldtype": "Float",
+			"width": 80
+		}, {
+			"label": _("Picked Qty"),
+			"fieldname": "actual_qty",
+			"fieldtype": "Float",
+			"width": 90
+		}, {
+			"label": _("Expected Arrival Data"),
```
Comment (szfisher) [inline erpnext/manufacturing/report/production_planning_report/production_planning_report.py]: Expected Arrival Date?

**Comment by rohitwaghchaure** [comment]
> Rohit, thank you for the report... please confirm the following:
> does the report show only work orders that are active or all?
> 
> Also, can we have a reverse report as well, where a raw material which could be missing in many items ... this will trigger a production when that raw material is received!

It will shows the all active work orders which has not completed and not stopped.
Where a raw material which could be missing in many items
Yes, we'll check for this

**Comment by rohitwaghchaure** [comment]
@toykraft thanks for the feedback, see my comments inline

Does the report show only work orders that are active or all?
**It will shows the all active work orders which has not completed and not stopped.**

Also, can we have a reverse report as well, where a raw material which could be missing in many items
**Yes, we'll check for this**

File: erpnext/manufacturing/report/production_planning_report/production_planning_report.py (hunk 1 above)
Comment (rohitwaghchaure) [reply erpnext/manufacturing/report/production_planning_report/production_planning_report.py]: This is the required by date mentioned in the purchase order item, this date helps user to know by when the raw materials will be expected in their warehouse

**Comment by rohitwaghchaure** [comment]
Raw materials allocation based on the highest sales order amount, this will be helpful for users to prioritize the orders

Screenshot 2020-03-13 at 6 12 39 PM

File: erpnext/manufacturing/report/production_planning_report/production_planning_report.py (hunk 1 above)
Comment (marination) [reply erpnext/manufacturing/report/production_planning_report/production_planning_report.py]:
```suggestion
			"label": _("Expected Arrival Date"),
```

Comment (rohitwaghchaure) [reply erpnext/manufacturing/report/production_planning_report/production_planning_report.py]: Fixed

File: erpnext/manufacturing/report/production_planning_report/production_planning_report.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+
+		for d in raw_materials:
+			if d.parent not in self.raw_materials_dict:
+				self.raw_materials_dict.setdefault(d.parent, [])
+
+			rows = self.raw_materials_dict[d.parent]
+			rows.append(d)
+
+	def get_item_details(self):
+		if not self.item_codes: return
```
Comment (marination) [inline erpnext/manufacturing/report/production_planning_report/production_planning_report.py]: `AttributeError: 'ProductionPlanReport' object has no attribute 'item_codes'` if nothing is fetched from `get_raw_materials`. Instantiate item_codes as an empty list in __init__ .

File: erpnext/manufacturing/report/production_planning_report/production_planning_report.py [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+				and d.remaining_qty != d.required_qty):
+				row = self.get_args()
+				d.warehouse = self.filters.raw_material_warehouse
+				d.required_qty = d.remaining_qty
+				d.alloted_qty = 0
+				row.update(d)
+				self.data.append(row)
+
+	def pick_materials_from_warehouses(self, args, order_data, warehouses):
+		required_qty = args.required_qty
```
Comment (marination) [inline erpnext/manufacturing/report/production_planning_report/production_planning_report.py]: `required_qty` isn't used anywhere else. this line can be removed.

File: erpnext/manufacturing/report/production_planning_report/production_planning_report.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+			"fieldtype": "Select",
+			"options": ["Delivery Date", "Total Amount"],
+			"default": "Delivery Date"
+		},
+		{
+			"fieldname":"include_subassembly_raw_materials",
+			"label": __("Include Sub-assembly Raw Materials"),
+			"fieldtype": "Check",
+			"default": 0
+		},
```
Comment (marination) [inline erpnext/manufacturing/report/production_planning_report/production_planning_report.js]: hide this if based on is work order using eval

File: erpnext/manufacturing/report/production_planning_report/production_planning_report.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+			"options": "Warehouse",
+			"get_query": function() {
+				var company = frappe.query_report.get_filter_value('company');
+				return {
+					filters: {
+						"company": company
+					}
+				};
+			}
+		},
```
Comment (marination) [inline erpnext/manufacturing/report/production_planning_report/production_planning_report.js:77]: hide this if based_on is work order using eval

File: erpnext/manufacturing/report/production_planning_report/production_planning_report.py [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+			"production_item_name": "",
+			"qty_to_manufacture": "",
+			"produced_qty": ""
+		})
+
+	def get_columns(self):
+		based_on = self.filters.based_on
+
+		self.columns = [{
+			"label": _("Id"),
```
Comment (marination) [inline erpnext/manufacturing/report/production_planning_report/production_planning_report.py]: Upper case 'ID'

File: erpnext/selling/doctype/sales_order_item/sales_order_item.json [hunk 7]
```diff
@@ -753,12 +756,24 @@
    "fieldname": "additional_notes",
    "fieldtype": "Text",
    "label": "Additional Notes"
+  },
+  {
+   "fieldname": "bom_no",
+   "fieldtype": "Link",
+   "label": "BOM No",
```
Comment (marination) [inline erpnext/selling/doctype/sales_order_item/sales_order_item.json:763]: get query for all BOM No fields in child table, filter on row item_code

File: erpnext/manufacturing/report/production_planning_report/production_planning_report.py [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+			"fieldname": "warehouse",
+			"fieldtype": "Link",
+			"width": 100
+		}, {
+			"label": _("Required Qty"),
+			"fieldname": "required_qty",
+			"fieldtype": "Float",
+			"width": 100
+		}, {
+			"label": _("Alloted Qty"),
```
Comment (marination) [inline erpnext/manufacturing/report/production_planning_report/production_planning_report.py]:
```suggestion
			"label": _("Allotted Qty"),
```

File: erpnext/manufacturing/report/bom_operations_time/bom_operations_time.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+	return frappe._dict({
+		"name": "",
+		"item": "",
+		"item_name": "",
+		"uom": ""
+	})
+
+def get_columns(filters):
+	return [{
+		"label": _("BOM Id"),
```
Comment (marination) [inline erpnext/manufacturing/report/bom_operations_time/bom_operations_time.py]:
```suggestion
		"label": _("BOM ID"),
```

**Review by marination — CHANGES_REQUESTED** [review]
In Production Planning for Sales Order, it is sorted by **delivery date** first and then **base grand total**. It's hard for a user to know that since there is no feedback. Maybe add these columns so the user can know how it is sorted.

**Comment by rohitwaghchaure** [comment]
Opened PR on develop branch https://github.com/frappe/erpnext/pull/21763

**Pull request closed without merging**